pkg/diagnostics/systemd/systemd.go

```diff
@@ -0,0 +1,547 @@
+package systemd
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"github.com/openshift/origin/pkg/diagnostics/log"
+	"github.com/openshift/origin/pkg/diagnostics/types"
+	"io"
+	"os/exec"
+	"regexp"
+)
+
+type logEntry struct {
+	Message string // I feel certain we will want more fields at some point
+}
+
+type logMatcher struct { // regex for scanning log messages and interpreting them when found
+	Regexp         *regexp.Regexp
+	Level          log.Level
+	Id             string
+	Interpretation string // log with above level+id if it's simple
+	KeepAfterMatch bool   // usually note only first matched entry, ignore rest
+	Interpret      func(  // run this for custom logic on match
+		env *types.Environment,
+		entry *logEntry,
+		matches []string,
+	) bool // KeepAfterMatch?
+}
+
+type unitSpec struct {
+	Name        string
+	StartMatch  *regexp.Regexp // regex to look for in log messages indicating startup
+	LogMatchers []logMatcher   // suspect log patterns to check for - checked in order
+}
+
+//
+// -------- Things that feed into the diagnostics definitions -----------
+// Search for Diagnostics for the actual diagnostics.
+
+// Reusable log matchers:
+var badImageTemplate = logMatcher{
+	Regexp: regexp.MustCompile(`Unable to find an image for .* due to an error processing the format: %!v\\(MISSING\\)`),
+	Level:  log.InfoLevel,
+	Interpretation: `
+This error indicates openshift was given the flag --images including an invalid format variable.
+Valid formats can include (literally) ${component} and ${version}.
+This could be a typo or you might be intending to hardcode something,
+such as a version which should be specified as e.g. v3.0, not ${v3.0}.
+Note that the --images flag may be supplied via the OpenShift master,
+node, or "openshift ex registry/router" invocations and should usually
+be the same for each.`,
+}
+
+// captures for logMatcher Interpret functions to store state between matches
+var tlsClientErrorSeen map[string]bool
+
+// Specify what units we can check and what to look for and say about it
+var unitLogSpecs = []*unitSpec{
```

Comment:
Sorry, I guess I commented the line which would not be fixed. Please delete the comment and close the code expansion.